### ISSUE TITLE
Feeding 2D tracks for reconstruction

### ISSUE THREAD

**Issue by jytime** [cvg/Hierarchical-Localization#322]
Hi Paul,

Thanks for your great work. It is really helpful.

I was wondering is it possible to use 2D tracks (probably computed by PixSfM or other algorithms) for reconstruction? I am trying to do this by modifying `extract_features.main` to `features_from_track` and `match_features.match_from_paths` to `match_from_paths_by_track`. 

Generally tracks is a tensor with the shape of `[B, N, 2]`, where B is the number of image frames, N is the number of tracks, 2 is xy. I tried to feed tracks whose average sub-pixel accuracy is around 50%, i.e., half of the points have < 1 pixel distance to the ground truth. However, the reconstruction result is not very satisfactory. On MegaDepth, its AUC10 is around 45% while SP&SG is around 65%. 

Do you think the less-than-ideal results stem from the inaccuracy of the tracks, or should I be looking into a more refined method of integrating them? Are there specific hyperparameters should be changed? For example, my matching scores come from cosine similarity which is usually >0.7, while I noticed the original matching scores may be quite small. Or generally, should we try to have more keypoints and matches, or filter the existing one as much as we can?

Thank you for your guidance and insights. 

Best Regards

```
def features_from_track(conf: Dict,
         image_dir: Path,
         export_dir: Optional[Path] = None,
         as_half: bool = True,
         image_list: Optional[Union[Path, List[str]]] = None,
         feature_path: Optional[Path] = None,
         overwrite: bool = False,
         tracks = None,
         confidence = None,
         des = None,) -> Path:
    logger.info('Extracting local features with configuration:'
                f'\n{pprint.pformat(conf)}')

    dataset = ImageDataset(image_dir, conf['preprocessing'], image_list)
    if feature_path is None:
        feature_path = Path(export_dir, conf['output']+'.h5')
    feature_path.parent.mkdir(exist_ok=True, parents=True)
    skip_names = set(list_h5_names(feature_path)
                     if feature_path.exists() and not overwrite else ())
    dataset.names = [n for n in dataset.names if n not in skip_names]
    if len(dataset.names) == 0:
        logger.info('Skipping the extraction.')
        return feature_path

    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    
    loader = torch.utils.data.DataLoader(
        dataset, num_workers=1, shuffle=False, pin_memory=True)
    
    for idx, data in enumerate(tqdm(loader)):
        name = dataset.names[idx]
        # I rename all the images by order 
        # and save them as 0000.png, 0001.png ...
        # so that I can easily get their original index here
        our_idx = int(os.path.basename(name)[:4])
        
        pred = {}
        # to colmap
        pred['keypoints'] = tracks[0,our_idx].cpu().numpy() + 0.5
        confidence[confidence<0.5] = 0
        
        pred['scores'] = confidence[0, our_idx,].cpu().numpy()
        pred['descriptors'] = des[our_idx].permute(1,0).cpu().numpy()
        uncertainty = confidence[0, our_idx,].mean().cpu().numpy() / 2
        pred['image_size'] = original_size = data['original_size'][0].numpy()
        
        if as_half:
            for k in pred:
                dt = pred[k].dtype
                if (dt == np.float32) and (dt != np.float16):
                    pred[k] = pred[k].astype(np.float16)

        with h5py.File(str(feature_path), 'a', libver='latest') as fd:
            try:
                if name in fd:
                    del fd[name]
                grp = fd.create_group(name)
                for k, v in pred.items():
                    grp.create_dataset(k, data=v)
                if 'keypoints' in pred:
                    grp['keypoints'].attrs['uncertainty'] = uncertainty
            except OSError as error:
                if 'No space left on device' in error.args[0]:
                    logger.error(
                        'Out of disk space: storing features on disk can take '
                        'significant space, did you enable the as_half flag?')
                    del grp, fd[name]
                raise error

        del pred

    logger.info('Finished exporting features.')
    return feature_path



@torch.no_grad()
def match_from_paths_by_track(conf: Dict,
                     pairs_path: Path,
                     match_path: Path,
                     feature_path_q: Path,
                     feature_path_ref: Path,
                     overwrite: bool = False,
                     tracks = None,
                     cos_simi = None,
                     des = None,) -> Path:
    logger.info('Matching local features with configuration:'
                f'\n{pprint.pformat(conf)}')

    if not feature_path_q.exists():
        raise FileNotFoundError(f'Query feature file {feature_path_q}.')
    if not feature_path_ref.exists():
        raise FileNotFoundError(f'Reference feature file {feature_path_ref}.')
    match_path.parent.mkdir(exist_ok=True, parents=True)

    assert pairs_path.exists(), pairs_path
    pairs = parse_retrieval(pairs_path)
    pairs = [(q, r) for q, rs in pairs.items() for r in rs]
    pairs = find_unique_new_pairs(pairs, None if overwrite else match_path)
    if len(pairs) == 0:
        logger.info('Skipping the matching.')
        return

    device = 'cuda' if torch.cuda.is_available() else 'cpu'

    dataset = FeaturePairsDataset(pairs, feature_path_q, feature_path_ref)
    loader = torch.utils.data.DataLoader(
        dataset, num_workers=5, batch_size=1, shuffle=False, pin_memory=True)
    writer_queue = WorkQueue(partial(writer_fn, match_path=match_path), 5)

    for idx, data in enumerate(tqdm(loader, smoothing=.1)):
        # Pick the original idx of each pair
        left_idx = int(os.path.basename(pairs[idx][0])[:4])
        right_idx = int(os.path.basename(pairs[idx][1])[:4])

        pair = names_to_pair(*pairs[idx])
        pred = {}
        
        # Tracks (B, N, 2) have already be sorted 
        # so their matches should be 0 - 0, 1 - 1, n - n
        pred['matches0'] = torch.arange(len(des)).to(device)
        pred['matches1'] = torch.arange(len(des)).to(device)

        # cos_simi (N, B, B) to (N, 1, 1)
        # cos_simi is cosine similarity, normalized to 0-1
        cross_conf = cos_simi[:,left_idx, right_idx]
        cross_conf = cross_conf * cross_conf
        invalid_mask = cross_conf < 0.75
        
        pred['matches0'][invalid_mask] = -1
        pred['matches1'][invalid_mask] = -1
        
        pred['matches0'] = pred['matches0'][None]
        pred['matches1'] = pred['matches1'][None]
        
        # the matching score is similarity
        cross_conf[invalid_mask] = 0
        pred['matching_scores0'] = cross_conf[None].to(device)
        pred['matching_scores1'] = cross_conf[None].to(device)

        writer_queue.put((pair, pred))
    writer_queue.join()
    logger.info('Finished exporting matches.')

```

**Comment by jytime** [comment]
Hi Paul,

Thanks for the quick response. Yes, I tried the track number of 256, 1024, and 4096, where 4096 shows the best result of AUC@10=45. 

The tracks are distributed over the entire image. Here is an example from IMC:

<img width="581" alt="Screenshot 2023-10-19 at 20 42 09" src="https://github.com/cvg/Hierarchical-Localization/assets/37647961/15ecd1f2-bb50-4bb4-a384-b477366e67b0">

And here is an sample to show the quality of tracks. The top is pred and the bottom is gt.

<img width="860" alt="Screenshot 2023-10-19 at 20 14 01" src="https://github.com/cvg/Hierarchical-Localization/assets/37647961/ae7d2488-4e91-4422-a958-e9106767eac6">


Its quality looks good to me and hence I was a bit confused about the COLMAP performance. I guess I need to double check some details :D

**Comment by jytime** [comment]
By the way, when the keypoints are defined in the original resolution, do we still need to specially deal with the keypoints scales? I noticed there is a step for keypoint scale in `extract_features.main`. I think this is because SP will resize the image to 1024, and we need to rescale the SP keypoints back to the original resolution, e.g., 512x512.  If tracks are already set in the 512x512 pixel coordinate, is scale adjustment still needed?

**Comment by sarlinpe** [comment]
> By the way, when the keypoints are defined in the original resolution, do we still need to specially deal with the keypoints scales?

No scaling is needed when the points are at the original image resolution.

> The top is pred and the bottom is gt.

What happens when you feed the ground truth tracks? are you sure that this ground truth is sub-pixel accurate? In phototourism datasets the baseline between images is rather small so having well-localized points is critical.

Since these tracks are so good, you could also disable the geometric verification with `hloc.reconstruction.main(..., skip_geometric_verification=True)`.

If this doesn't help, the next steps is to check how many of these ground truth tracks survive the reconstruction process by inspecting the number of points reconstructed by each image in the final 3D model. Looking at statistics printed by COLMAP during incremental mapping would also be useful.
